### ISSUE TITLE
Segmentation fault hazard in Vulnerability Detector

### ISSUE THREAD

**Issue by vikman90** [wazuh/wazuh#8324]
|Wazuh version|Component|Install type|
|---|---|---|
| 3.11.0 | Vulnerability Detector | Manager |

From version 3.11 Vulnerability Detector improved the matching of the Windows OS CPE (ea50647). Function `wm_vuldet_adapt_os_cpe()` assumes that its input is non-null. However,  `wm_vuldet_decode_cpe()`, that produces the CPE string, may return `NULL` ([wm_vuln_detector_nvd.c:396](https://github.com/wazuh/wazuh/blob/ecc19c3a5735509370c96b1b818c0c5d67658cd0/src/wazuh_modules/vulnerability_detector/wm_vuln_detector_nvd.c#L393-L397)).

## Segfault report

We've found no current case that makes `wm_vuldet_decode_cpe()` return NULL. But, if we force it to return a null pointer, this is the segmentation fault reported by GDB:

```
Thread 8 "wazuh-modulesd" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff4df9640 (LWP 104971)]
0x00005555556046a8 in wm_vuldet_adapt_os_cpe (ag_cpe=0x0) at wazuh_modules/vulnerability_detector/wm_vuln_detector_nvd.c:3904
3904	    if (*ag_cpe->part != 'o') {
```

**Comment by katebee** [comment]
We have also been seeing segfaults in `wazuh-modulesd` since ~10:54 GMT (Wazuh v4.0.4 running on Ubuntu)
Disabling the vulnerability-detector seems to be a viable workaround so expect we have the same issue.

**Comment by delray25** [comment]
I have also seen segfaults in `wazuh-modulesd` since yesterday.  It happens after downloading/processing the vulnerability file from nist. See the debug log:

`2021/04/22 09:23:03 wazuh-modulesd[5269] url.c:346 at wurl_request_uncompress_bz2_gz(): DEBUG: File from URL 'https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-2011.json.gz' was successfully uncompressed into 'tmp/vuln-temp-fitted'`
`2021/04/22 09:23:08 wazuh-modulesd:vulnerability-detector[5269] wm_vuln_detector.c:3722 at wm_vuldet_index_feed(): DEBUG: (5414): Refreshing 'National Vulnerability Database' databases.`
`2021/04/22 09:23:08 wazuh-modulesd:vulnerability-detector[5269] wm_vuln_detector.c:2467 at wm_vuldet_insert(): DEBUG: (5415): Inserting vulnerabilities.`
`2021/04/22 09:23:08 wazuh-modulesd:vulnerability-detector[5269] wm_vuln_detector.c:2502 at wm_vuldet_insert(): DEBUG: (5419): Inserting NVD vulnerabilities section.`

**Kapow!!**

Below the kernel log:  (from a different session)
`Apr 22 05:55:39 UbuntuWazuh kernel: [58651.705772] wazuh-modulesd[7946]: segfault at 0 ip 0000000000460961 sp 00007fbbd6e2aa30 error 4 in wazuh-modulesd[400000+e8000]`

I worked around it by setting enabled to 'no' in `ossec.conf` as below:

```
<!-- Aggregate vulnerabilities -->
<provider name="nvd">
    <enabled>no</enabled>
    <update_from_year>2010</update_from_year>
    <update_interval>1h</update_interval>
</provider>
```

**Comment by vikman90** [comment]
Hi everyone,

We were reported about the crash you're mentioning. We fixed it yesterday (https://github.com/wazuh/wazuh/commit/4cbd1e85eeee0eb0d8247fa7228f590a9dd24153), and also released [Wazuh 4.1.5](https://github.com/wazuh/wazuh/releases/tag/v4.1.5) which applies this patch.

Please upgrade your manager to 4.1.5 to get this problem solved. Also, we are working on a backport of the fix for version 3.13.

On the other hand, we just discovered this other bug reviewing the first patch. This could make Modulesd crash if the NVD feed contained an invalid CPE string. I managed to cause this failure by simulating an invalid CPE, but this has not happened with any current NVD feed. That's why I mentioned that this is a hazard. However, we will fix this problem ASAP in order to prevent an invalid feed from making Modulesd crash again.

Thank you for commenting.
Best regards.

**Comment by kurianoff** [comment]
@vikman90 Please advise if the fix for 3.13 was also implemented. We are using AWS Elasticsearch which keeps us with the old version of Kibana (7.9.1), so having a fix for Wazuh 3.13 is very beneficial for us, otherwise we're missing on vulnerability db updates. Thanks a lot in advance!!!

**Comment by vikman90** [comment]
@kurianoff yes, we released [v3.13.3](https://github.com/wazuh/wazuh/releases/tag/v3.13.3) applying this fix last week. Hope that helps you.

Best regards!

**Comment by akukhar-qualtie** [comment]
@vikman90 should we wait for updated [wazuh/wazuh-kibana-app](https://github.com/wazuh/wazuh-kibana-app) v3.13.3 version? 
Thanks.

**Comment by vikman90** [comment]
Hi @akukhar-qualtie. Wazuh App for Kibana has been already released for [Elastic 7.9.2](https://github.com/wazuh/wazuh-kibana-app/releases/tag/v3.13.3-7.9.2) and [Elastic 7.8.0](https://github.com/wazuh/wazuh-kibana-app/releases/tag/v3.13.3-7.8.0).

You can reach the full list of packages in the documentation of [3.13 / Packages list](https://documentation.wazuh.com/3.13/installation-guide/packages-list/index.html). Here are the [release notes](https://documentation.wazuh.com/3.13/release-notes/release_3_13_3.html).

Best regards.
